### ISSUE TITLE
More mixins

### ISSUE THREAD

**Issue by hameerabbasi** [numpy/numpy#10446]
I was considering more mixins like `NDArrayOperatorsMixin`. Some I was considering were:

- `NDArrayUfuncsMixin` for things like `duckarray.sin()`, `duckarray.exp()`, etc, which would defer to `np.sin` and `np.exp`.
- `NDArrayReductionsMixin` for `ufunc`-based reductions like `duckarray.sum(axis=1)`, etc which would defer to `np.add.reduce`.

I'm willing to put this together, but a question I have is, is there a better way than just listing out all the options?

**Comment by njsmith** [comment]
Do we need separate mixins for these things?

**Comment by hameerabbasi** [comment]
It's worth discussing whether to put everything into a single mixin, `NDArrayMixin`, which would work if anyone implements `__array_ufunc__`.

**Comment by shoyer** [comment]
Can you explain what these specific mixins would use for their implementation? An example would go a long ways. 

NDArrayOperatorsMixin exists to let you use NumPy's `__array_ufunc__` protocol to implement Python operators. We don't need that for ufuncs because `__array_ufunc__` directly serves the purpose of letting you define implementations of `np.sin()`.

Reductions might make more sense, but many reductions (e.g., mean) are not actually implemented as a single ufunc reduce method. So we would need to define a new protocol for those first, e.g., `__array_aggregate__`.

**Comment by eric-wieser** [comment]
> Can you explain what these specific mixins would use for their implementation?

I would speculate something as simple as:
```
class NDArrayMethodMixin:
    mean = np.mean
    sum = np.sum
    prod = np.prod
    # etc
```

Which provided all the method versions of the free functions. It's not clear to me that we should be encouraging this, but it means that users can write `a.mean()` and it will work for duck-types, rather than having to remember to write `np.mean(a)` or similar

**Comment by shoyer** [comment]
@eric-wieser `np.mean(a)` already calls `a.mean()` if it exists, so that mixin example would lead to infinite recursion.

**Comment by hameerabbasi** [comment]
I wasn't considering implementing `mean` at this time, more like just things of the form `duckarray.ufunc()`, `duckarray.ufunc(other)` and `duckarray.reduce()`

For example,
```
class NDArrayMixin:
    def exp(self, out=None):
        return np.exp(self, out=out)

    def sum(self, axis=None, dtype=None, out=None, keepdims=False):
        return np.add.reduce(self, axis=axis, dtype=dtype, out=out, keepdims=keepdims)
```

Mean can be easily modified to return, for example, `sum / shape` of some sort.

**Comment by eric-wieser** [comment]
Why would you want to implement `duckarray.ufunc_name(other)` when `ndarray.ufunc_name(other)` doesn't exist?

**Comment by njsmith** [comment]
Basically my feeling is, it would be great to make `NDArrayOperatorsMixin` implement more of the methods that `np.ndarray` provides, however makes sense to do that. I don't think we need other mixins, or that we should add new methods that aren't on `np.ndarray`, because the point of `NDArrayOperatorsMixin` is to make it easier to implement the `np.ndarray` interface.

**Comment by hameerabbasi** [comment]
> Why would you want to implement `duckarray.ufunc_name(other)` when `ndarray.ufunc_name(other)` doesn't exist? Neither `ndarray.sin` nor `ndarray.exp` exist.

It doesn't? I must have assumed as I used it with other libraries, my mistake. In that case, I withdraw the proposal for ufuncs and only keep the reduction proposal.

**Comment by shoyer** [comment]
I actually think more mixins could potentially make sense, for distinct
parts of the ndarray interface. Not every duck array needs to support
arithmetic (e.g., consider arrays stored on disk with h5py), and not every
object that could implement arithmetic with __array_ufunc__ satisfies the
full duck array interface (e.g., consider xarray.Dataset).

**Comment by shoyer** [comment]
Which specific reductions/aggregations would we implement on a mixin?

I suppose most of these are probably defined in [`numpy.core._methods`](https://github.com/numpy/numpy/blob/master/numpy/core/_methods.py):
- min, max, any, all, sum and prod directly use a ufunc reduce method.
- mean, var and std have more complicated implantations (see that file), but are small wrappers of ufuncs.
- cumsum and cumprod seem like they could logically use the ufunc accumulate method, but their implementation is elsewhere in C (possibly using the ufunc underneath, I haven't checked yet)

**Comment by hameerabbasi** [comment]
> min, max, any, all, sum and prod directly use a ufunc reduce method.

These, definitely.

> mean, var and std have more complicated implantations (see that file), but are small wrappers of ufuncs.

I still agree we should implement these. If we're implementing separate mixins, these would also need operators, right? If it's one mixin I'm all for it. If separate... Then where do we put them?

> cumsum and cumprod seem like they could logically use the ufunc accumulate method, but their implementation is elsewhere in C (possibly using the ufunc underneath, I haven't checked yet)

This would need to be checked.

**Comment by njsmith** [comment]
> Not every duck array needs to support arithmetic (e.g., consider arrays stored on disk with h5py) and not every object that could implement arithmetic with `__array_ufunc__` satisfies the full duck array interface (e.g., consider `xarray.Dataset`).

Oo, examples, lets dig into the details :-). I don't understand the first one – why would we want `h5py_arr + 1` to fail? Why is `h5py` implementing `__array_ufunc__` if it doesn't want to support arithmetic? Or are you imagining that there would be operations in `NDArrayOperatorsMixin` that don't reduce to `__array_ufunc__` calls, and if so, which ones are you thinking of?

For `xarray.Dataset` similarly, can you give some some examples of operations that might go into `NDArrayOperatorsMixin` but which `xarray.Dataset` doesn't want to support?

**Comment by shoyer** [comment]
h5py.Dataset stores data, but doesn't support any computation directly.
They can be bigger than what fits in memory, so you don't want to
implicitly convert full arrays into NumPy (e.g., Dask might be more
appropriate). For these sorts of reasons, I'm pretty sure h5py doesn't
implement __array_ufunc__. But it does implement other ndarray
methods/properties like indexing, dtype and shape. Some of these might be
appropriate for mixins, e.g., an ndim property defined by shape, or nbytes
defined by shape and dtype.

**Comment by shoyer** [comment]
> I may be being very ignorant here, but wouldn't any array with a dtype have that already, since we're not overloading dtype but rather using Numpy's own?

Yes, most arrays implementing `__array_ufunc__` probably also implement a `dtype` property. Here's the problem:
- Usually `dtype` is an actual `numpy.dtype` instance.
- `numpy.dtype.type` returns a numpy scalar type, e.g., `np.float64`.
- NumPy scalar types always return numpy arrays (or subclasses) of the appropriate type when called on an object. I don't think we have any way to override this behavior (yet).

Let me illustrate the problem with `pandas.Series`, which implements most of the appropriate methods for overloading NumPy functions:
```
In [27]: import numpy as np

In [28]: import pandas as pd

In [29]: np.dtype(np.float64).type is np.float64
Out[29]: True

In [30]: np.float64(pd.Series([1, 2]))
Out[30]: array([ 1.,  2.])
# result is a base numpy array, not a pandas.Series
```

**Comment by shoyer** [comment]
One way to resolve this issue would be to add some sort of dispatching on scalar types to call an `astype()` method, if available, instead of casting inputs to a NumPy array. But whereas the methods that are implemented directly as a ufunc can be override simply with `__array_ufunc__`, for `mean()` we would now need:
- `__array_ufunc__` support for `reduce`
- `__array_ufunc__` support for arithmetic (probably OK)
- `dtype` attribute that is a numpy dtype.
- `ndim` and `shape` attributes (for counting the number of elements).
- `astype` to support casting with scalar types.

All of this is reasonable in principle but the interface is starting to get quite a bit messier, with more design considerations/requirements for the mixin.

**Comment by hameerabbasi** [comment]
> Perhaps a `no_scalar=False` argument added to `ufunc.*`, which prevents this decay?

This would be generally useful, but is probably a separate issue. But as pointed out, we probably don't need it for `mean`.

**Comment by eric-wieser** [comment]
> If you're implementing your own __array_ufunc__, you can chose to let 0-dimensional results decay to scalars or not based on your implementation. I don't think we need a separate keyword argument for that.

Not reliably, because when you call the ufunc on a base ndarray, there's nothing you can do to stop it decaying. This argument would be especially useful within `__array_ufunc__`

**Comment by shoyer** [comment]
> Not reliably, because when you call the ufunc on a base ndarray, there's nothing you can do to stop it decaying. This argument would be especially useful within __array_ufunc__.

Yes, you would have to add some logic in your wrapper function. That's ugly, but not too bad to write.

**Comment by hameerabbasi** [comment]
> Not reliably, because when you call the ufunc on a base ndarray, there's nothing you can do to stop it decaying (and it's sometimes impossible to tell whether it did, for object arrays). This argument would be especially useful within `__array_ufunc__`

+1, I propose we make `astype` a `ufunc`, add that argument. It will be better in the long term, and easier for downstream developers.

**Comment by shoyer** [comment]
> +1, I propose we make astype a ufunc, add that argument. It will be better in the long term, and easier for downstream developers.

All ufuncs already have `dtype` and `casting` arguments in their `__call__` method. So it would probably make more sense to add an `identity` ufunc for this rather than `astype`. (Also note the different default values for some of the optional parameters on `astype` vs ufuncs.)

In practice `np.identity()` could reuse loops from `np.positive()`, but we would need to add support for non-numeric dtypes.

**Comment by hameerabbasi** [comment]
In any case, since `mean` and `np.identity` could take a while, do I have a green light to open a PR with the reductions in https://github.com/numpy/numpy/issues/10446#issuecomment-359498049 ?

**Comment by shoyer** [comment]
Actually, `np.positive` would probably suffice for the purposes of type casting in `np.mean`. There is no need for supporting non-numeric types in mean. 

I would be okay getting started with less, but it would be best to clearly set expectations for mixin users now, as much as is possible. Otherwise people might start using the mixin even if future methods like mean would not be appropriate.

**Comment by hameerabbasi** [comment]
I ran a quick test...

```
import numpy as np
np.mean([False, True])
Out[3]: 0.5
np.sum([False, True])
Out[5]: 1
np.positive(np.sum([False, True]), dtype=np.float16)
Out[6]: 1.0
np.positive(np.sum([False, True]) / 2, dtype=np.float16)
Out[7]: 0.5
```

So we're good. However, @eric-wieser's comment about object arrays stands. That could be filed as a separate issue to be fixed in the future.

I would suggest putting `reduce`/`accumulate`/simple `ufunc` combos in separate mixins anyway, for exactly the reason you're concerned about.

**Comment by hameerabbasi** [comment]
Gentle bump to facilitate discussion. I'll edit #10460 to match consensus here. :-)

**Comment by hameerabbasi** [comment]
I was coming back to this as it's needed in a number of duck-array projects. 🙂 

We can implement almost all array methods and properties using `__array_function__` and `__array_ufunc__`. It would tremendously speed up writing a duck array class if this was implemented, and simplify code in other cases where the class has already been written.
